File: cache/src/main/java/com/mm/http/CacheCallAdapter.java
package com.mm.http;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import javax.annotation.Nullable;

import retrofit2.Call;

/**
 * Created by : majian
 * Date : 2021/8/24
 */

public interface CacheCallAdapter<R, T> {
    Type responseType();

    T adapt(Call<R> call);

    okhttp3.Call rawCall(Class<?> service, Method method, Object[] args);

    abstract class Factory {
        /**
         * Returns a call adapter for interface methods that return {@code returnType}, or null if it
         * cannot be handled by this factory.
         */
        public abstract @Nullable
        CacheCallAdapter<?, ?> get(Type returnType, Annotation[] annotations, RetrofitCache retrofit);

        /**
         * Extract the upper bound of the generic parameter at {@code index} from {@code type}. For
         * example, index 1 of {@code Map<String, ? extends Runnable>} returns {@code Runnable}.
         */
        protected static Type getParameterUpperBound(int index, ParameterizedType type) {
            return Utils.getParameterUpperBound(index, type);
        }

        /**
         * Extract the raw class type from {@code type}. For example, the type representing {@code
         * List<? extends Runnable>} returns {@code List.class}.
         */
        protected static Class<?> getRawType(Type type) {
            return Utils.getRawType(type);
        }
    }

}

File: cache/src/main/java/com/mm/http/CacheCallAdapter.kt
package com.mm.http

import retrofit2.Call
import java.lang.reflect.Method
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * Created by : majian
 * Date : 2021/8/24
 */
interface CacheCallAdapter<R, T> {

    fun responseType(): Type

    fun adapt(call: Call<R>): T

    fun rawCall(service: Class<*>, method: Method, args: Array<Any>?): okhttp3.Call

    abstract class Factory {
        /**
         * Returns a call adapter for interface methods that return `returnType`, or null if it
         * cannot be handled by this factory.
         */
        abstract operator fun get(returnType: Type, annotations: Array<Annotation>, retrofit: RetrofitCache): CacheCallAdapter<*, *>?

        companion object {
            /**
             * Extract the upper bound of the generic parameter at `index` from `type`. For
             * example, index 1 of `Map<String, ? extends Runnable>` returns `Runnable`.
             */
            @JvmStatic
            protected fun getParameterUpperBound(index: Int, type: ParameterizedType): Type {
                return Utils.getParameterUpperBound(index, type)
            }

            /**
             * Extract the raw class type from `type`. For example, the type representing `List<? extends Runnable>` returns `List.class`.
             */
            @JvmStatic
            protected fun getRawType(type: Type): Class<*> {
                return Utils.getRawType(type)
            }
        }
    }
}